dashboard/modules/job/job_agent.py

```diff
@@ -0,0 +1,376 @@
+          - The basename of driver filename without extension in the job
+          package archive.
+        JAVA:
+          - The driver class full name in the job package archive.
+        """
+        return self._job_info["driverEntry"]
+
+    def driver_args(self):
+        """ The driver arguments in list.
+        PYTHON:
+          -  The arguments to pass to the main() function in driver entry.
+          e.g. [1, False, 3.14, "abc"]
+        JAVA:
+          - The arguments to pass to the driver command line.
+          e.g. ["-custom-arg", "abc"]
+        """
+        driver_args = self._job_info.get("driverArgs", [])
+        assert isinstance(driver_args, list)
+        return driver_args
+
+    def set_driver(self, driver: asyncio.subprocess.Process):
+        """ Set the driver process instance.
+        """
+        self._driver = driver
+
+    def driver(self):
+        return self._driver
+
+    def set_initialize_task(self, task):
+        """ Set an async task that initializes the job environment.
+        """
+        self._initialize_task = task
+
+    def initialize_task(self):
+        return self._initialize_task
+
+    def env(self):
+        """ The environment vars to pass to job config.
+        """
+        job_package_dir = job_consts.JOB_UNPACK_DIR.format(
+            temp_dir=self._temp_dir, job_id=self._job_id)
+        env_dict = {"RAY_JOB_DIR": job_package_dir}
+        # Support json values for env.
+        for k, v in self._job_info.get("env", {}).items():
+            if isinstance(v, str):
+                env_dict[k] = v
+            else:
+                env_dict[k] = json.dumps(v)
+        return env_dict
+
+
+class JobProcessor:
+    def __init__(self, job_info):
+        assert isinstance(job_info, JobInfo)
+        self._job_info = job_info
+
+    async def _download_package(self, http_session, url, filename):
+        job_id = self._job_info.job_id()
+        cmd_index = self._get_next_cmd_index()
+        logger.info("[%s] Start download[%s] %s to %s", job_id, cmd_index, url,
+                    filename)
+        async with http_session.get(url, ssl=False) as response:
+            with open(filename, "wb") as f:
+                while True:
+                    chunk = await response.content.read(
+                        job_consts.DOWNLOAD_BUFFER_SIZE)
+                    if not chunk:
+                        break
+                    f.write(chunk)
+            logger.info("[%s] Finished download[%s] %s to %s", job_id,
+                        cmd_index, url, filename)
+
+    async def _unpack_package(self, filename, path):
+        code = f"import shutil; " \
+               f"shutil.unpack_archive({repr(filename)}, {repr(path)})"
+        unzip_cmd = [self._get_current_python(), "-c", code]
+        await self._check_call_cmd(unzip_cmd)
+
+    @staticmethod
+    def _get_next_cmd_index(start_index=[1]):
+        cmd_index = start_index[0]
+        start_index[0] += 1
```

Comment:
why not just use a global variable? it took me a while to understand what this does.

Reply:
That requires a global variable or a class member. I will refactor the code to make it easier to understand.